subprojects/core/src/main/java/org/gradle/initialization/properties/DefaultSystemPropertiesInstaller.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.initialization.properties;
+
+import org.gradle.api.Project;
+import org.gradle.api.internal.GradleInternal;
+import org.gradle.api.internal.StartParameterInternal;
+import org.gradle.initialization.EnvironmentChangeTracker;
+
+import java.util.Map;
+
+import static org.gradle.internal.Cast.uncheckedNonnullCast;
+
+public class DefaultSystemPropertiesInstaller implements SystemPropertiesInstaller {
+
+    private final EnvironmentChangeTracker environmentChangeTracker;
+
+    private final StartParameterInternal startParameter;
+
+    private final GradleInternal gradleInternal;
+
+    public DefaultSystemPropertiesInstaller(
+        EnvironmentChangeTracker environmentChangeTracker,
+        StartParameterInternal startParameter,
+        GradleInternal gradleInternal
+    ) {
+        this.environmentChangeTracker = environmentChangeTracker;
+        this.startParameter = startParameter;
+        this.gradleInternal = gradleInternal;
+    }
+
+    @Override
+    public void setSystemPropertiesFrom(GradlePropertiesInternal gradleProperties) {
+        // TODO:configuration-cache What happens when a system property is set from a Gradle property and
+        //    that same system property is then used to set a Gradle property from an included build?
+        //    e.g., included-build/gradle.properties << systemProp.org.gradle.project.fromSystemProp=42
+        setSystemPropertiesFromGradleProperties(gradleProperties.getDefaultProperties());
+        setSystemPropertiesFromGradleProperties(gradleProperties.getOverrideProperties());
```

Comment:
Can we just do `setSystemPropertiesFromGradleProperties(gradleProperties.mergeProperties(emptyMap()))` instead of these two calls? The last one wins anyway. I also vote for introducing `gradleProperties.getProperties` instead of the monstrosity above :) - this would be the third occurrence.